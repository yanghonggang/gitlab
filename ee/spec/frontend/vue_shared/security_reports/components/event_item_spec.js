import { GlButton } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import Component from 'ee/vue_shared/security_reports/components/event_item.vue';
import NoteHeader from '~/notes/components/note_header.vue';

describe('Event Item', () => {
  let wrapper;

  const mountComponent = (options, mountFn = shallowMount) => {
    wrapper = mountFn(Component, options);
  };

  const noteHeader = () => wrapper.find(NoteHeader);

  describe('initial state', () => {
    const propsData = {
      id: 123,
      createdAt: 'createdAt',
      headerMessage: 'header message',
      author: {
        name: 'Tanuki',
        username: 'gitlab',
      },
    };

    const slots = { default: '<p>Test</p>' };

    beforeEach(() => {
      mountComponent({ propsData, slots });
    });

    afterEach(() => {
      wrapper.destroy();
    });

    it('passes the expected values to the note header component', () => {
      expect(noteHeader().props()).toMatchObject({
        noteId: propsData.id,
        author: propsData.author,
        createdAt: propsData.createdAt,
        showSpinner: false,
      });
    });

    it('uses the fallback icon', () => {
      expect(wrapper.props().iconName).toBe('plus');
    });

    it('uses the fallback icon class', () => {
      expect(wrapper.props().iconClass).toBe('ci-status-icon-success');
    });

    it('renders the action buttons container', () => {
      expect(wrapper.find('.action-buttons')).toExist();
    });

    it('renders the default slot', () => {
      expect(wrapper.html()).toEqual(expect.stringContaining('<p>Test</p>'));
    });
  });
  describe('with action buttons', () => {
    const propsData = {
      author: {
        name: 'Tanuki',
        username: 'gitlab',
      },
      actionButtons: [
        {
          iconName: 'pencil',
          onClick: jest.fn(),
          title: 'Foo Action',
        },
        {
          iconName: 'remove',
          onClick: jest.fn(),
          title: 'Bar Action',
        },
      ],
    };

    afterEach(() => {
      wrapper.destroy();
    });

    beforeEach(() => {
      mountComponent({ propsData }, mount);
    });

    it('renders the action buttons container', () => {
      expect(wrapper.find('.action-buttons')).toExist();
    });

    it('renders the action buttons', () => {
      expect(wrapper.findAll(GlButton)).toHaveLength(2);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('emits the button events when clicked', () => {
      const buttons = wrapper.findAll(GlButton);
      buttons.at(0).trigger('click');
      return wrapper.vm
        .$nextTick()
        .then(() => {
          buttons.at(1).trigger('click');
          return wrapper.vm.$nextTick();
        })
        .then(() => {
          expect(propsData.actionButtons[0].onClick).toHaveBeenCalledTimes(1);
          expect(propsData.actionButtons[1].onClick).toHaveBeenCalledTimes(1);
        });
    });
  });
});
