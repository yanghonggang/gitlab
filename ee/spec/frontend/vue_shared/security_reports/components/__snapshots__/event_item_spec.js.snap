// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Event Item with action buttons renders the action buttons 1`] = `
<div
  class="d-flex align-items-center"
>
  <div
    class="circle-icon-container gl-flex-shrink-0 gl-align-self-start ci-status-icon-success"
  >
    <svg
      class="gl-icon s16"
      data-testid="plus-icon"
    >
      <use
        href="#plus"
      />
    </svg>
  </div>
   
  <div
    class="ml-3 flex-grow-1"
    data-qa-selector="event_item_content"
  >
    <div
      class="note-header-info pb-0"
    >
      <!---->
       
      <a
        class="author-name-link js-user-link"
        data-username="gitlab"
      >
         
        <span
          class="note-header-author-name gl-font-weight-bold"
        >
          Tanuki
        </span>
      </a>
       
      <!---->
       
      <span
        class="text-nowrap author-username"
      >
        <a
          class="author-username-link"
        >
          <span
            class="note-headline-light"
          >
            @gitlab
          </span>
        </a>
         
        <!---->
      </span>
       
      <span
        class="note-headline-light note-headline-meta"
      >
        <span
          class="system-note-message"
          data-qa-selector="system_note_content"
        >
          Â·
        </span>
         
        <!---->
         
        <!---->
          
        <!---->
      </span>
    </div>
     
  </div>
   
  <div
    class="gl-flex-shrink-0 gl-align-self-start"
  >
    <button
      class="btn btn-default btn-md gl-button btn-default-tertiary btn-icon"
      title="Foo Action"
      type="button"
    >
      <!---->
       
      <svg
        class="gl-button-icon gl-icon s16"
        data-testid="pencil-icon"
      >
        <use
          href="#pencil"
        />
      </svg>
        
      <!---->
    </button>
    <button
      class="btn btn-default btn-md gl-button btn-default-tertiary btn-icon"
      title="Bar Action"
      type="button"
    >
      <!---->
       
      <svg
        class="gl-button-icon gl-icon s16"
        data-testid="remove-icon"
      >
        <use
          href="#remove"
        />
      </svg>
        
      <!---->
    </button>
  </div>
</div>
`;
